### PR TITLE
Fix transition between ball placement and free kick

### LINKED ISSUE #3237: Fix transiton from ball placement to direct free
### Acceptance criteria

- [ ]

### PR DESCRIPTION
**Pull request by williamckha**
### Description

Fixes broken transitions between SetPlays in `PlaySelectionFSM` (notably, ball placement to free kick). This PR adds an internal transition for `SetPlay_S` and updates `setupSetPlay` to only construct a new SetPlay if it is not already the current play.

### Testing Done

Added test in `play_selection_fsm_test.cpp`

### Resolved Issues

Resolves #3237 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/play_selection_fsm.h [hunk 1]
```diff
@@ -110,14 +108,32 @@ struct PlaySelectionFSM
 
             // Check for transitions to other states, if not then default to running the
             // current play
-            SetPlay_S + Update_E[gameStateHalted_G] / setupHaltPlay_A     = Halt_S,
-            SetPlay_S + Update_E[gameStateStopped_G] / setupStopPlay_A    = Stop_S,
-            SetPlay_S + Update_E[gameStatePlaying_G] / setupOffensePlay_A = Playing_S,
+            SetPlay_S + Update_E[gameStateHalted_G] / (resetSetPlay_A, setupHaltPlay_A) =
+                Halt_S,
```
Comment (nimazareian) [inline src/software/ai/play_selection_fsm.h:112]: Didn't know this was possible with boost SML. does this call both actions?

**Review by nimazareian — DISMISSED** [review]
Just tried it with auto ref and it seemed to work well! :100:

File: src/software/ai/play_selection_fsm.h (hunk 1 above)
Comment (williamckha) [reply src/software/ai/play_selection_fsm.h:112]: Yep, and I think in that order as well

**Review by nimazareian — APPROVED** [review]

**Review by itsarune — APPROVED** [review]
nice work!

File: src/software/ai/play_selection_fsm.cpp [hunk 2]
```diff
@@ -41,45 +41,70 @@ bool PlaySelectionFSM::gameStateSetupRestart(const Update& event)
...
 {
-    current_play.reset();
     if (event.game_state.isOurBallPlacement())
     {
-        event.set_current_play(std::make_unique<BallPlacementPlay>(ai_config));
+        if (current_set_play != TbotsProto::PlayName::BallPlacementPlay)
+        {
+            current_set_play = TbotsProto::PlayName::BallPlacementPlay;
+            event.set_current_play(std::make_unique<BallPlacementPlay>(ai_config));
+        }
```
Comment (itsarune) [inline src/software/ai/play_selection_fsm.cpp:50]: nit: you could probably merge it with the `if` conditional

`if (event.game_state.isOurBallPlacement() && current_set_play != TbotsProto::PlayName::BallPlacementPlay)`

**Comment by itsarune** [comment]
feel free to merge. my nit isn't blocking

File: src/software/ai/play_selection_fsm.cpp (hunk 2 above)
Comment (williamckha) [reply src/software/ai/play_selection_fsm.cpp:50]: If `event.game_state.isOurBallPlacement()` is true, we don't want to enter any of the other `else if` branches if `current_set_play == TbotsProto::PlayName::BallPlacementPlay)`

**Pull request merged**